core/src/sumeragi/main_loop.rs

```diff
@@ -430,325 +449,381 @@ impl Sumeragi {
                 {
                     // NOTE: Up until this point it was unknown which block is expected to be received,
                     // therefore all the signatures (of any hash) were collected and will now be pruned
-                    add_signatures::<false>(&mut new_block, voting_signatures.drain(..));
-                    *voting_block = Some(new_block);
+                    add_signatures::<false>(
+                        &mut valid_block,
+                        voting_signatures.drain(..),
+                        &self.topology,
+                    );
+
+                    *voting_block = self.try_commit_block(valid_block, is_genesis_peer);
+                } else {
+                    // FIX: Restore `voting_block`
+                    // Currently it's not possible because block writer needs to be released
                 }
             }
-            (BlockMessage::BlockSigned(BlockSigned { hash, signatures }), Role::ProxyTail) => {
-                trace!(block_hash=%hash, "Received block signatures");
+            (BlockMessage::BlockSigned(BlockSigned { signatures }), Role::ProxyTail) => {
+                info!(
+                    peer_id=%self.peer_id,
+                    role=%self.role(),
+                    "Received block signatures"
+                );
 
-                let roles: &[Role] = if current_view_change_index >= 1 {
+                let roles: &[Role] = if view_change_index >= 1 {
                     &[Role::ValidatingPeer, Role::ObservingPeer]
                 } else {
                     &[Role::ValidatingPeer]
                 };
-                let valid_signatures =
-                    current_topology.filter_signatures_by_roles(roles, &signatures);
-
-                if let Some(voted_block) = voting_block.as_mut() {
-                    let voting_block_hash = voted_block.block.as_ref().hash_of_payload();
-
-                    if hash == voting_block_hash {
-                        add_signatures::<true>(voted_block, valid_signatures);
-                    } else {
-                        debug!(%voting_block_hash, "Received signatures are not for the current block");
-                    }
+                let valid_signatures = self
+                    .topology
+                    .filter_signatures_by_roles(roles, &signatures)
+                    .cloned();
+
+                if let Some(mut voted_block) = voting_block.take() {
+                    add_signatures::<true>(&mut voted_block, valid_signatures, &self.topology);
+                    *voting_block = self.try_commit_block(voted_block, is_genesis_peer);
                 } else {
                     // NOTE: Due to the nature of distributed systems, signatures can sometimes be received before
                     // the block (sent by the leader). Collect the signatures and wait for the block to be received
                     voting_signatures.extend(valid_signatures);
                 }
             }
```

Comment:
By looking at this code and `try_commit` it look like we don't check that signature indeed belong to the peer. 
So any peer (including malicious leader itself) can send `BlockSigned` message with random block mathing signatures for enough validating/obsering peers and block will be committed. 

Another problem is that on this path signatures are not checked for duplication.
So as before peer can submit `BlockSigned` message with the same index and signature and block will be committed.

Reply:
1. `ValidBlock` is guaranteed to have all valid signatures except for the `ProxyTail` signature
2. `valid_bloc.commit` will check that `ProxyTail` signature is valid
3. `try_commit` is only invoked by the `ProxyTail`. You can see there is an `assert` statement inside

Reply:
I see no problem with peer-signature correspondence, but `let valid_signatures` would be premature naming since the validity is guaranteed in `add_signatures`

I agree that signatures can duplicate

Reply:
fixed